### ISSUE TITLE
[...array] of an annotated parameter with a default value in async function shows a warning

### ISSUE THREAD

**Issue by phistuck** [google/closure-compiler#3009]
The following code -
```js
/** @param {!Array<string>=} foo */
async function baz(foo = [])
{
  const newFoo = [...foo];
  return newFoo;
}
```
Results in this warning -
```text
JSC_TYPE_MISMATCH: actual parameter 1 of $jscomp.arrayFromIterable does not match formal parameter
found   : (Array<string>|undefined)
required: (Arguments<string>|Array<string>|Iterable<string>|string) at line 4 character 21
  const newFoo = [...foo];
```

(Making the parameter non-optional by removing `=` does not change anything)

https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%2540language_out%2520ES5%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F**%2520%2540param%2520%257B!Array%253Cstring%253E%253D%257D%2520foo%2520*%252F%250Aasync%2520function%2520baz(foo%2520%253D%2520%255B%255D)%250A%257B%250A%2520%2520const%2520newFoo%2520%253D%2520%255B...foo%255D%253B%250A%2520%2520return%2520newFoo%253B%250A%257D

**Comment by lauraharker** [comment]
Thanks for the report.

This is a byproduct of how we currently typecheck default parameters and async functions:
we transpile them before typechecking, so the typechecker sees transpiled code. Something like:
``` javascript
/** @param {!Array<string>=} foo */
function baz(foo){
  foo = foo === undefined ? [] : foo;
  var newFoo;
  return $jscomp.asyncExecutePromiseGeneratorProgram(function($jscomp$generator$context) {
    newFoo=[].concat($jscomp.arrayFromIterable(foo));
    return $jscomp$generator$context.return(newFoo)
  });
};
```

Our typechecker doesn't realize that in the transpiled code, `foo` will never be undefined where it's actually used.

This will be fixed once we stop transpiling default parameters before typechecking. hopefully by the end of July.

**Comment by lauraharker** [comment]
Fixed as of v20181125
